internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -230,6 +230,18 @@ func (r *Reconciler) reconcileControlPlane(ctx context.Context, s *scope.Scope)
 		return err
 	}
 
+	// If the controlPlane has infrastructureMachines and the InfrastructureMachineTemplate has changed on this reconcile
+	// delete the old template.
+	// This is a best effort deletion only and may leak templates if an error occurs during reconciliation.
+	// TODO:(killianmuldoon) Make a more holistic template deletion flow.
+	if s.Blueprint.HasControlPlaneInfrastructureMachine() && s.Current.ControlPlane.InfrastructureMachineTemplate != nil {
+		if s.Current.ControlPlane.InfrastructureMachineTemplate.GetName() != s.Desired.ControlPlane.InfrastructureMachineTemplate.GetName() {
+			if err := r.Client.Delete(ctx, s.Current.ControlPlane.InfrastructureMachineTemplate); err != nil {
+				return errors.Wrapf(err, "failed to delete old template %s", tlog.KObj{Obj: s.Desired.ControlPlane.Object})
```

Comment:
```suggestion
				return errors.Wrapf(err, "failed to delete old template %s", tlog.KObj{Obj: s.Current.ControlPlane.InfrastructureMachineTemplate})
```
Not sure what the intention is but I think this error should contain the name of the template we couldn't delete, if we also want to have the control plane in the error message I think we should add both.

=> I think both would be ideal